AJ-1283: Add unit tests to `Data.js`

jladieu opened the pull request:
### Jira Ticket: https://broadworkbench.atlassian.net/browse/AJ-1283

### Dependencies
Depends on https://github.com/DataBiosphere/terra-ui/pull/4157 for common test fixture data.

## Summary of changes:
This will provide a baseline to act as guardrails for changes I'll be making to the status polling.

### What
- Adds tests in prep for making changes to `Data.js`

### Why
- `Data.js` is large and unwieldy and lacks unit tests; this baseline will make it easier to add more tests going forward, and provide some basic protection for some changes targeted to the status polling.

### Testing strategy
- Add new tests to establish baseline expected behavior.

jladieu committed:
AJ-1283: Add unit tests to `Data.js` to provide guardrails for upcoming changes to status polling.

jladieu committed:
Drop the extra internal `async` call from the tests.

nawatts commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -1532,3 +1532,8 @@ export const navPaths = [
 ];
+
+// These modules are exported for testing only.
+export const testOnlyExports = {
```
For application code, I don't think we care enough about curating the exports / interface to distinguish these as test only. I would vote to simply export the component where it's defined.

```js
export const WorkspaceData = ...
```

nawatts commented on src/pages/workspaces/workspace/Data.test.ts:
```diff
@@ -0,0 +1,137 @@
+    };
+
+    const mockGetSchema = jest.fn().mockResolvedValue([]);
+    const mockAjax: DeepPartial<AjaxContract> = {
+      Workspaces: {
+        workspace: (_namespace, _name) => ({
+          details: jest.fn().mockResolvedValue(workspace),
+          listSnapshots: jest.fn().mockResolvedValue([]),
+          entityMetadata: jest.fn().mockResolvedValue({}),
+        }),
+      } as DeepPartial<WorkspacesContract>,
+      WorkspaceData: {
+        getSchema: mockGetSchema,
+      } as DeepPartial<WorkspaceDataContract>,
+      Apps: {
+        listAppsV2: jest.fn().mockResolvedValue([{ ...wdsApp, status }]),
+      } as DeepPartial<AppsContract>,
+    };
```
The `DeepPartial` type should handle validating that nested objects conform to the type. There's no need to cast them.

```suggestion
    const mockAjax: DeepPartial<AjaxContract> = {
      Workspaces: {
        workspace: (_namespace, _name) => ({
          details: jest.fn().mockResolvedValue(workspace),
          listSnapshots: jest.fn().mockResolvedValue([]),
          entityMetadata: jest.fn().mockResolvedValue({}),
        }),
      },
      WorkspaceData: {
        getSchema: mockGetSchema,
      },
      Apps: {
        listAppsV2: jest.fn().mockResolvedValue([{ ...wdsApp, status }]),
      },
    };
```

jladieu replied:
Oh good call!  This also let me drop the extra type declarations of the inner contracts.

nawatts commented on src/pages/workspaces/workspace/Data.test.ts:
```diff
@@ -0,0 +1,137 @@
+      workspace,
+      refreshWorkspace,
+      storageDetails,
+    } as WorkspaceDataProps;
```
```suggestion
    const workspaceDataProps: WorkspaceDataProps = {
      namespace,
      name,
      workspace,
      refreshWorkspace,
      storageDetails,
    };
```

Declaring the type for the variable provides more type safety than casting.

nawatts commented on src/pages/workspaces/workspace/Data.test.ts:
```diff
@@ -0,0 +1,137 @@
+  }
+
+  // SIFERS setup, see: https://medium.com/@kolodny/testing-with-sifers-c9d6bb5b362
+  function setup({
```
It would be nice to declare types for this function's parameters and return value.

```ts
type SetupOptions = { ... };

type SetupResult = { ... };

function setup(opts: SetupOptions): SetupResult { ... }
```

jladieu replied:
I like it.  I added these types in to make `setup` more explicit.  I put the type declarations _inside_ the `describe` as I thought they should be scoped _just_ to this set of tests on the premise that other eventual tests in this suite might have different `setup`.  LMK if this goes against the spirit of your other suggestion to bump the `Ajax` and `WorkspaceDataProps` up to the top level, or if it makes sense to keep the types local where possible.

nawatts replied:
Fair enough. I don't feel very strongly about the location of types. Since Ajax is "global" for this module, I think it makes sense to move the Ajax type up, but keeping the others here makes sense.

nawatts commented on src/pages/workspaces/workspace/Data.test.ts:
```diff
@@ -0,0 +1,137 @@
+    const { workspaceDataProps } = setup({ workspace: defaultAzureWorkspace, status: 'PROVISIONING' });
+
+    // Act
+    await act(() => {
+      render(h(WorkspaceData, workspaceDataProps));
+    });
```
I'm surprised nothing complains about this. `act` is typed as returning a Promise only if its given a function that returns a Promise.

It seems like this suggestion is unnecessary, but it's in line with existing code.

```suggestion
    await act(async () => {
      render(h(WorkspaceData, workspaceDataProps));
    });
```

jladieu replied:
Hah...I don't really grok `async` / `await` yet, but was trying to get the tersest invocation possible and the test suite and linter didn't yell at me for stripping these out.  I added this back in though for consistency.

nawatts commented on src/pages/workspaces/workspace/Data.test.ts:
```diff
@@ -0,0 +1,137 @@
+    });
+
+    // Assert
+    expect(screen.getByText(/Data tables are unavailable/)).toBeVisible();
```
```suggestion
    expect(screen.getByText('Data tables are unavailable')).toBeVisible();
```

Style nit: suggest using a string unless a regex is necessary.

jladieu replied:
I landed on regex because using the string I needed to include the entire text of the DOM element to get a hit -- this kinda felt brittle to me.  I didn't want a simple text/grammar change/addition/clarification to cause a _definite_ test failure.    Happy to go full string though if, even with this tradeoff in mind, you'd still recommend string.

nawatts replied:
That makes sense.

nawatts commented on src/pages/workspaces/workspace/Data.test.ts:
```diff
@@ -0,0 +1,137 @@
+describe('WorkspaceData', () => {
+  // When Data.js is broken apart and the WorkspaceData component is converted to TypeScript,
+  // this type belongs there.
+  interface WorkspaceDataProps {
```
Style nit: suggest moving this type and AjaxContract to the module level instead of inside the `describe`.

jladieu replied:
Done.  Although I added two more types for the `SetupOptions` and `SetupResult` right in this spot.  LMK if they should get the same treatment.

nawatts reviewed (changes requested):
Thanks for daring to break ground on a test for this component!

Requesting changes to remove casts as they reduce type safety. I'd also prefer to update the export style of WorkspaceData so as to not introduce a new pattern.

jladieu committed:
Incorporate `@nawatts` feedback:

* Re-inline `testOnlyExports`
* Move top-level `WorkspaceDataProps` and `AjaxContract` types to module level.
* Declare types for `SetupOptions` and `SetupResult`
* Eliminate unnecessary casting to `DeepPartial` (as well as the type declarations).
* Declare type instead of casting where possible.
* Re-introduce `async()` to `act` invocations for consistency.

jladieu commented on src/pages/workspaces/workspace/Data.test.ts:
```diff
@@ -28,31 +26,49 @@ jest.mock('src/libs/ajax', (): AjaxExports => {
-  // When Data.js is broken apart and the WorkspaceData component is converted to TypeScript,
-  // this type belongs there.
-  interface WorkspaceDataProps {
+  type SetupOptions = {
```
@nawatts at your suggestion, I added these types in for `Setup`, but I have two questions:

* does it make sense to keep them local to the `describe`?  I thought other test cases might have their own version of `setup` and therefore these belonged here rather than at the module level
* Just curious about the use of terms `Options` vs. `Props` -- later on I have a type called `WorkspaceDataProps` which could also, in my mind, just be named `WorkspaceDataOptions` -- are there conventions on when one name is used vs. the other?

nawatts replied:
> does it make sense to keep them local to the describe?

Yes, I think so. I think my comments about the scope for type declarations were tied to thinking of the common case where a module exports one thing and its test has one describe block (a pattern that not all modules fit into).

> Just curious about the use of terms Options vs. Props

"Props" is a React term. https://react.dev/learn#sharing-data-between-components

So the convention is that a React function component takes one argument: a "props" object.

Vs "options" is a more generic term for functions in general.
https://www.codereadability.com/what-are-javascript-options-objects

jladieu commented on src/pages/workspaces/workspace/Data.test.ts:
```diff
@@ -28,31 +26,49 @@ jest.mock('src/libs/ajax', (): AjaxExports => {
-  function setup({
-    namespace = 'test-namespace',
-    name = 'test-name',
-    workspace,
```
I deliberately left `workspace` unset in this old style, so that the client tests would _have_ to provide a `workspace` value.  However, by introducing the `SetupOptions` type, I couldn't figure out how to leave `workspace` unset in the invocation of `setup`, which meant adding `...defaultSetupOptions` and sprinkling it throughout.

Is there an idiomatic way for me to preserve the old invocation style, where the individual tests _had_ to provide a workspace, but didn't have to provide an entire `SetupOptions` class?

```setup({ workspace: defaultAzureWorkspace, status: 'RUNNING' })```

instead of:
```setup({ ...defaultSetupOptions, workspace: defaultAzureWorkspace, status: 'RUNNING' })```

nawatts replied:
You could mark the property as optional in the Options type.
https://www.typescriptlang.org/docs/handbook/2/objects.html#optional-properties

Or possibly have the setup function take an argument that's a `Partial<WorkspaceDataProps>` and merge it with the default values.
https://www.typescriptlang.org/docs/handbook/utility-types.html#partialtype

jladieu replied:
I dug up an old [resource](https://bobbyhadz.com/blog/typescript-object-parameter-default-value#:~:text=To%20set%20default%20value%20for,all%20its%20properties%20to%20optional) to figure out what I wanted to accomplish here, and pushed up this [attempt](https://github.com/DataBiosphere/terra-ui/pull/4169/commits/d3dec1ee23877714e504f5ba5425573531dac93c) to smooth over this rough edge.

jladieu committed:
Improve invocation pattern for `setup()`

This makes it so tests _must_ provide `workspace` and `status` by making them required fields on `SetupOptions`.  Everything else, while defined as optional, will be assigned a default value by the `setup` method if omitted.

This also allows `defaultSetupOptions` to be dropped.
